Akpm find closest lib

sopyb opened the pull request.

sopyb committed:
lib: move find_closest() and find_closest_descending() to lib functions

Move the utility macros find_closest() and find_closest_descending()
from inline macros to proper library functions in lib/.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

commodo commented on lib/find_closest.c:
```diff
@@ -0,0 +1,72 @@
+/*
+ * Find closest element functions
+ *
+ * Copyright (C) 2025 Alexandru Soponar <asoponar@taladin.ro>
```
No idea if we can use copy-rights here.
This isn't new code.

commodo commented on lib/find_closest.c:
```diff
@@ -0,0 +1,72 @@
+{
+	unsigned int i;
+	unsigned int array_size = as - 1;
+	long mid_x, left, right;
```
maybe use `s32` instead of `long`?  for consistency

commodo commented on include/linux/find_closest.h:
```diff
@@ -0,0 +1,40 @@
+
+#include <linux/types.h>
+
+/**
```
i think doc-strings need to be in one place; not both;

since they are in the `find_closes.c`, i think these could be removed;

commodo commented on include/linux/util_macros.h:
```diff
@@ -45,35 +29,41 @@
+}
 
-#endif
+#endif /* _LINUX_HELPER_MACROS_H_ */
```
this is patch-noise;
remove it;

commodo commented on include/linux/util_macros.h:
```diff
@@ -45,35 +29,41 @@
-#define is_insidevar(ptr, var)						\
-	((uintptr_t)(ptr) >= (uintptr_t)(var) &&			\
-	 (uintptr_t)(ptr) <  (uintptr_t)(var) + sizeof(var))
+#define is_insidevar(ptr, var, ...)					\
```
this is also patch-noise
remove it;

if wanting to fix/address this, then it should be a separate patch;

commodo commented on include/linux/util_macros.h:
```diff
@@ -45,35 +29,41 @@
-	(__fc_i);							\
-})
+#define find_closest_descending(x, a, as) \
+	find_closest_descending_s32((s32)(x), (const s32 *)(a), (as))
```
explicit type-casting in the kernel is discouraged, unless it's unavoidable;
no idea why;
maybe because it can hide weird type-casting bugs;

commodo commented on lib/find_closest.c:
```diff
@@ -0,0 +1,72 @@
+ * Find closest element functions
+ *
+ * Copyright (C) 2025 Alexandru Soponar <asoponar@taladin.ro>
+ * Based on util_macros.h implementation by Alexandru Ardelean <aardelean@baylibre.com>
```
maybe let's not add this information here (for various reasons);

part of it is: this was initially written by someone else, then fixed by me;
the code is now a mix of authors;

sopyb replied:
> maybe let's not add this information here (for various reasons);
> 
> part of it is: this was initially written by someone else, then fixed by me; the code is now a mix of authors;

I wasn't aware this was originally written by someone else think the right thing to do would be to replace line with something along the lines of:
 * Based on util_macros.h implementation

Alongside removing the copyright notice.

sopyb replied:
I'll leave this open instead of marking it as resolved to hear if you think I should proceed otherwise.

commodo commented on include/linux/find_closest.h:
```diff
@@ -0,0 +1,40 @@
+ * Note: This function is specifically designed for s32 (signed 32-bit) arrays
+ * to properly handle negative and positive values.
+ */
+unsigned int find_closest_s32(s32 x, const s32 *a, unsigned int as);
```
the `find_closest_s32()` naming is interesting;
initially, i thought maybe about doing it as just `find_closest()`, but i can see it being useful to have it as `find_closest_s32()`;

no idea what upstream will say;
let's leave it as-is for now;

as context: sometimes future-proofing is discouraged, especially in cases where there is no clear idea if there will ever be a `find_closest_u32`  or `find_closest_s64` version;  in the kernel, things are completely self-contained/isolated, so if there is no clear idea if something will ever be used in the future, it is discouraged and the remark will be "we can change it later if we need to"

commodo commented:
Please run (in the repo):
`./scripts/checkpatch.pl --strict -g HEAD~10..` and address any complaints that script has.

`-g HEAD~10..` is the commit range, so feel free to adjust that as needed.

sopyb commented:
> Please run (in the repo): `./scripts/checkpatch.pl --strict -g HEAD~10..` and address any complaints that script has.

I've addressed all the style issues reported by checkpatch.pl. The only remaining warning is about needing to update MAINTAINERS for the new file, but this isn't necessary in this case since the new file would naturally fall under Andrew Morton's maintenance as part of the LIBRARY CODE section.

commodo commented on include/linux/util_macros.h:
```diff
@@ -15,24 +16,7 @@
-})
+#define find_closest(x, a, as)	find_closest_s32(x, a, as)
 
 /**
```
you can also remove these doc-strings;

commodo commented on include/linux/util_macros.h:
```diff
@@ -3,6 +3,7 @@
+#include <linux/find_closest.h>
 
 /**
  * find_closest - locate the closest element in a sorted array
```
you can also remove these doc-strings;

commodo commented on lib/find_closest.c:
```diff
@@ -0,0 +1,71 @@
+ *
+ * Similar to find_closest_s32() but 'a' is expected to be sorted in descending
+ * order. The iteration is done in reverse order, so that the comparison
+ * of 'right' & 'left' also works for unsigned numbers.
```
this comment is no longer needed `The iteration is done in reverse order, so that the comparison
 * of 'right' & 'left' also works for unsigned numbers.`
 
the way that this works now, is that `find_closest_{descending_}s32()` is enforcing types to `s32`
previously (as it was a macro) it wasn't doing that;

sopyb replied:
you're right it slipped past me as I was copying over the doc-strings

commodo commented:
> > Please run (in the repo): `./scripts/checkpatch.pl --strict -g HEAD~10..` and address any complaints that script has.
> 
> I've addressed all the style issues reported by checkpatch.pl. The only remaining warning is about needing to update MAINTAINERS for the new file, but this isn't necessary in this case since the new file would naturally fall under Andrew Morton's maintenance as part of the LIBRARY CODE section.

yeah, no need for MAINTATINERS file update;

commodo commented:
did you try to run the kunit test
IIRC, it used to be something like running    `  ./tools/testing/kunit/kunit.py run "*util_macros*"`

sopyb commented:
> did you try to run the kunit test IIRC, it used to be something like running ` ./tools/testing/kunit/kunit.py run "*util_macros*"`

Yeah I did it's all passing.

commodo commented on drivers/hwmon/ina2xx.c:
```diff
@@ -213,7 +213,7 @@ static const struct ina2xx_config ina2xx_config[] = {
  * https://www.ti.com/lit/ds/symlink/ina226.pdf.
  */
-static const int ina226_avg_tab[] = { 1, 4, 16, 64, 128, 256, 512, 1024 };
+static const s32 ina226_avg_tab[] = { 1, 4, 16, 64, 128, 256, 512, 1024 };
```
these `int` -> `s32` conversions are incomplete;
i.e.: there are still places where `int` is still used for the array of `s32` values;
either all places get converted from `int` -> `s32` or we remove them from the current patches;

but `u16` -> `s32` or `u32` -> `s32` are mandatory now (since the compiler should complain about incompatible pointer types);

commodo commented:
`drivers/iio/adc/mcp3564.c` needs to be patched;
```
static const unsigned int mcp3564_oversampling_avail[] = {
........
};  
```

commodo commented:
`drivers/iio/adc/ad7606.c` also needs a bit of work;

there are multiple `unsigned int` types passed to `find_closest()`

commodo commented:
`drivers/hwmon/pmbus/max20730.c` needs addressing

commodo commented:
`drivers/hwmon/max6639.c` needs addressing;

commodo commented:
`drivers/hwmon/ltc4282.c`  needs addressing

commodo commented:
`drivers/hwmon/lm95234.c` needs addressing

commodo commented:
`drivers/hwmon/lm75.c`  needs addressing

commodo commented:
`drivers/leds/rgb/leds-mt6370-rgb.c`  needs addressing

commodo commented:
`drivers/iio/light/max44009.c`  needs addressing

commodo commented:
`drivers/hwmon/w83795.c` needs addressing

commodo commented:
`drivers/hwmon/max1619.c`  needs addressing

sopyb committed:
hwmon: w83795: Fix type incompatibility with non-macro find_closest

The pwm_freq_cksel0 array was previously declared as u16 but used with
find_closest_descending(). Now that find_closest_descending() is
implemented as a function with 'int' parameters instead of a macro,
passing unsigned arrays causes type incompatibility errors. This patch
changes the array type from u16 to int to match the function's signature
and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: emc1403: Fix type incompatibility with non-macro find_closest

The ina3221_conv_time array was previously declared as u16 but used with
find_closest_descending(). With find_closest_descending() now implemented
as a function taking 'int' parameters instead of a macro, passing unsigned
arrays causes type incompatibility errors. Change the array type from u16
to int to ensure compatibility with the new function signature.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: ina3221: Fix type incompatibility with non-macro find_closest

The ina3221_conv_time array was previously declared as u16 but used with
find_closest(). With find_closest() now implemented as a function taking
signed int parameters instead of a macro, passing unsigned arrays causes
type incompatibility errors. This patch changes the array type from
u16 to int to ensure compatibility with the function signature and
prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: lm95234: Fix type incompatibility with non-macro find_closest

The update_intervals array was previously declared as u16 but used with
find_closest(). With find_closest() now implemented as a function taking
signed int parameters instead of a macro, passing unsigned arrays causes
type incompatibility errors. This patch changes the array type from
u16 to int to ensure compatibility with the function signature and
prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: max1619: Fix type incompatibility with non-macro find_closest

The update_intervals array was previously declared as u16 but used with
find_closest_descending(). With find_closest_descending() now implemented
as a function taking signed int parameters instead of a macro, passing
unsigned arrays causes type incompatibility errors. This patch changes
the array type from u16 to int to ensure compatibility with the function
signature and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

commodo commented on lib/find_closest.c:
```diff
@@ -0,0 +1,71 @@
+ */
+unsigned int find_closest_descending(int x, const int *a, unsigned int as)
+{
+	unsigned int i;
```
nitpick: there's this thing called reverse xmas tree which kernel people seem to like
```
	unsigned int array_size = as - 1;
	int mid_x, left, right;
	unsigned int i;
```

commodo commented on lib/find_closest.c:
```diff
@@ -0,0 +1,71 @@
+{
+	unsigned int i;
+	unsigned int array_size = as - 1;
+	int mid_x, left, right;
```
same here, reverse xmas tree

commodo commented:
a bunch of files from other subsystems were missed?
so far, these look like they're only from hwmon;
but there should be a few from iio and a bunch of others;

sopyb commented:
> a bunch of files from other subsystems were missed?
> so far, these look like they're only from hwmon;
> but there should be a few from iio and a bunch of others;

I checked users of check_closest, and these were the only cases that used unsigned integer arrays and needed to be changed to signed. That's why the changes are limited to hwmon. I built the kernel with `make allyesconfig` and didn't encounter any errors or warnings.

sopyb committed:
hwmon: lm75: Fix type incompatibility with non-macro find_closest

The default_sample_time and sample_times members of lm75_params were
previously declared as unsigned int but used withfind_closest(). With
find_closest() now implemented as a function taking signed int parameters
instead of a macro, passing unsigned arrays causes type incompatibility
errors. This patch changes the types from unsigned int and unsigned int*
to int respectively int* to ensure compatibility with the function
signature and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: ltc4282: Fix type incompatibility with non-macro find_closest

The ltc4282_out_rates array was previously declared as unsigned int but
used with find_closest(). With find_closest() now implemented as a function
taking signed int parameters instead of a macro, passing unsigned arrays
causes type incompatibility errors. This patch changes the array type from
unsigned int to int to ensure compatibility with the function signature and
prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: max6639: Fix type incompatibility with non-macro find_closest

The freq_table array was previously declared as unsigned int but used
with find_closest(). With find_closest() now implemented as a function
taking signed int parameters instead of a macro, passing unsigned arrays
causes type incompatibility errors. This patch changes the array type from
unsigned int to int to ensure compatibility with the function signature and
prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
hwmon: max20740: Fix type incompatibility with non-macro find_closest

The max_current array was previously declared as u32 but used with
find_closest(). With find_closest() now implemented as a function taking
signed int parameters instead of a macro, passing unsigned arrays causes
type incompatibility errors. This patch changes the array type from u32
to int to ensure compatibility with the function signature and prevent
compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
iio: ad7606: Fix type incompatibility with non-macro find_closest

The ad7606_oversampling_avail and ad7616_oversampling_avail arrays were
previously declared as unsigned int but used with find_closest(). With
find_closest() now implemented as a function taking signed int parameters
instead of a macro, passing unsigned arrays causes type incompatibility
errors. This patch changes the arrays type from unsigned int to int to
ensure compatibility with the function signature and prevent compilation
errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
iio: mcp3564: Fix type incompatibility with non-macro find_closest

The mcp3564_oversampling_avail array was previously declared as unsigned
int but used with find_closest(). With find_closest() now implemented as
a function taking signed int parameters instead of a macro,  passing
unsigned arrays causes type incompatibility errors. This patch changes the
arrays type from unsigned int to int to ensure compatibility with the
function signature and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
iio: max44009: Fix type incompatibility with non-macro find_closest

The max44009_int_time_ns_array array was previously declared as u32 but
used with find_closest(). With find_closest() now implemented as a
function taking signed int parameters instead of a macro, passing unsigned
arrays causes type incompatibility errors. This patch changes the arrays
type from u32 to int to ensure compatibility with the function signature
and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
leds: eds-mt6370-rgb: Fix type incompatibility with find_closest()

The common_tfreqs and mt6372_tfreqs arrays were previously declared as
unsigned int but used with find_closest(), which now takes signed int
parameters. Change these arrays from unsigned int to int to maintain type
compatibility with the find_closest() function signature and prevent
compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
regulator: max77857: Fix type incompatibility with find_closest()

The max77857_switch_freq array was previously declared as unsigned int
but used with find_closest(), which takes signed int parameters. Change
this array from unsigned int to int to maintain type compatibility with
the find_closest() function signature and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

sopyb committed:
watchdog: simatic-ipc-wdt: Fix type incompatibility with find_closest()

The wd_timeout_table array was previously declared as unsigned int
but used with find_closest(), which takes signed int parameters. Change
this array from unsigned int to int to maintain type compatibility with
the find_closest() function signature and prevent compilation errors.

Signed-off-by: Alexandru Soponar <asoponar@taladin.ro>

commodo commented on drivers/hwmon/ina3221.c:
```diff
@@ -204,8 +204,8 @@ static inline u32 ina3221_reg_to_interval_us(u16 config)
-	u32 vbus_ct = ina3221_conv_time[vbus_ct_idx];
-	u32 vsh_ct = ina3221_conv_time[vsh_ct_idx];
+	int vbus_ct = ina3221_conv_time[vbus_ct_idx];
+	int vsh_ct = ina3221_conv_time[vsh_ct_idx];
```
not sure if these should be changed;
i guess it's fine to keep; let's see what upstream says;
int/u32 convert a bit more seemlessly in C (vs C++); there is a lesser chance of bugs (than with pointer type conversions);
